File: backend/app/repositories/booking_repository.py
from typing import List, Optional
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from fastapi import HTTPException

from ..models.booking import BookingCreate
from ..db.models import Booking as BookingModel, User, Slot

class BookingRepository:
    def create_booking(self, db: Session, booking: BookingCreate) -> BookingModel:
        """
        Create a new booking in the database
        Raises IntegrityError if a duplicate booking is detected
        """
        # Check if user exists
        user = db.query(User).filter(User.id == booking.user_id).first()
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        # Check if slot exists
        slot = db.query(Slot).filter(Slot.id == booking.slot_id).first()
        if not slot:
            raise HTTPException(status_code=404, detail="Slot not found")

        # Check if slot is in the future
        if booking.booking_date < datetime.now():
            raise HTTPException(status_code=400, detail="Cannot book a slot in the past")

        # Create new booking
        try:
            db_booking = BookingModel(
                user_id=booking.user_id,
                slot_id=booking.slot_id,
                booking_date=booking.booking_date,
                status="confirmed"
            )
            db.add(db_booking)
            db.commit()
            db.refresh(db_booking)
            return db_booking
        except IntegrityError:
            db.rollback()
            raise HTTPException(
                status_code=400,
                detail="Duplicate booking: You already have a booking for this slot"
            )

    def get_user_bookings(self, db: Session, user_id: int) -> List[BookingModel]:
        """Get all bookings for a specific user"""
        return db.query(BookingModel).filter(BookingModel.user_id == user_id).all()

    def get_booking_by_id(self, db: Session, booking_id: int) -> Optional[BookingModel]:
        """Get a booking by its ID"""
        return db.query(BookingModel).filter(BookingModel.id == booking_id).first()

    def cancel_booking(self, db: Session, booking_id: int) -> Optional[BookingModel]:
        """Cancel a booking by changing its status"""
        booking = self.get_booking_by_id(db, booking_id)
        if booking:
            booking.status = "cancelled"
            db.commit()
            db.refresh(booking)
        return booking

File: backend/src/repositories/BookingRepository.js
const { db } = require('../database');
const Booking = require('../models/Booking');
const { logger } = require('../utils/logger');

class BookingRepository {
  constructor() {
    this.tableName = 'bookings';
  }

  /**
   * Find booking by ID
   * @param {string} id Booking ID
   * @returns {Promise<Booking|null>} Booking instance or null if not found
   */
  async findById(id) {
    try {
      if (!id) throw new Error('Booking ID is required');

      const result = await db(this.tableName)
        .where({ id })
        .first();

      return result ? new Booking(result) : null;
    } catch (error) {
      logger.error(`BookingRepository.findById error: ${error.message}`, { id });
      throw new Error(`Failed to find booking: ${error.message}`);
    }
  }

  /**
   * Find all bookings for a user
   * @param {string} userId User ID
   * @returns {Promise<Booking[]>} Array of bookings
   */
  async findByUserId(userId) {
    try {
      if (!userId) throw new Error('User ID is required');

      const results = await db(this.tableName)
        .where({ user_id: userId })
        .orderBy('booking_date', 'desc');

      return results.map(result => new Booking(result));
    } catch (error) {
      logger.error(`BookingRepository.findByUserId error: ${error.message}`, { userId });
      throw new Error(`Failed to find bookings for user: ${error.message}`);
    }
  }

  /**
   * Check if a slot is already booked
   * @param {string} slotId Slot ID
   * @param {Date} date Booking date
   * @returns {Promise<boolean>} True if slot is already booked
   */
  async isSlotBooked(slotId, date) {
    try {
      if (!slotId) throw new Error('Slot ID is required');

      const result = await db(this.tableName)
        .where({
          slot_id: slotId,
          booking_date: date
        })
        .whereNot({ status: 'cancelled' })
        .first();

      return !!result;
    } catch (error) {
      logger.error(`BookingRepository.isSlotBooked error: ${error.message}`, { slotId, date });
      throw new Error(`Failed to check if slot is booked: ${error.message}`);
    }
  }

  /**
   * Create a new booking
   * @param {Booking} booking Booking instance
   * @returns {Promise<Booking>} Created booking
   */
  async create(booking) {
    try {
      if (!(booking instanceof Booking)) {
        throw new Error('Invalid booking object');
      }

      const isBooked = await this.isSlotBooked(booking.slotId, booking.bookingDate);
      if (isBooked) {
        throw new Error('This slot is already booked');
      }

      const [result] = await db(this.tableName)
        .insert(booking.toDatabase())
        .returning('*');

      logger.info(`Created new booking with ID: ${result.id}`);
      return new Booking(result);
    } catch (error) {
      logger.error(`BookingRepository.create error: ${error.message}`, { booking });
      throw new Error(`Failed to create booking: ${error.message}`);
    }
  }

  /**
   * Update a booking
   * @param {string} id Booking ID
   * @param {Object} data Booking data to update
   * @returns {Promise<Booking>} Updated booking
   */
  async update(id, data) {
    try {
      if (!id) throw new Error('Booking ID is required');

      const booking = await this.findById(id);
      if (!booking) {
        throw new Error(`Booking not found with ID: ${id}`);
      }

      // Update booking properties
      Object.assign(booking, data);
      booking.validate();

      const [result] = await db(this.tableName)
        .where({ id })
        .update(booking.toDatabase())
        .returning('*');

      logger.info(`Updated booking with ID: ${result.id}`);
      return new Booking(result);
    } catch (error) {
      logger.error(`BookingRepository.update error: ${error.message}`, { id, data });
      throw new Error(`Failed to update booking: ${error.message}`);
    }
  }

  /**
   * Delete a booking
   * @param {string} id Booking ID
   * @returns {Promise<boolean>} True if booking was deleted
   */
  async delete(id) {
    try {
      if (!id) throw new Error('Booking ID is required');

      const result = await db(this.tableName)
        .where({ id })
        .delete();

      logger.info(`Deleted booking with ID: ${id}`);
      return result > 0;
    } catch (error) {
      logger.error(`BookingRepository.delete error: ${error.message}`, { id });
      throw new Error(`Failed to delete booking: ${error.message}`);
    }
  }
}

module.exports = new BookingRepository();
